### DISCUSSION TITLE
Extract Position Information of Keys / Values

### DISCUSSION THREAD

**Discussion by grahamc** [nlohmann/json#3103] [Q&A] [answered]
Hello everyone,

Is there a way to ask the library where in a document a key was defined?

I'm a user of Nix, which has a language builtin for parsing json. This builtin uses nlohmann/json for its implementation. The Nix language tries to track the position where values are defined, for example:

```nix
{
  foo = "Bar";
  baz.tux = {
    example = "hello";
  };
}
```

```
nix-repl> x = import ./test.nix 

nix-repl> builtins.unsafeGetAttrPos "foo" x       
{ column = 3; file = "/home/grahamc/test.nix"; line = 2; }

nix-repl> builtins.unsafeGetAttrPos "baz" x  
{ column = 3; file = "/home/grahamc/test.nix"; line = 3; }

nix-repl> builtins.unsafeGetAttrPos "tux" x.baz 
{ column = 3; file = "/home/grahamc/test.nix"; line = 3; }

nix-repl> builtins.unsafeGetAttrPos "example" x.baz.tux 
{ column = 5; file = "/home/grahamc/test.nix"; line = 4; }
```

I would like to extend this support to JSON documents, too.

For example, given a Nix expression like:

```nix
{
  json = builtins.fromJSON ''
{
  "foo": "bar"
}
'';
}
```

today Nix doesn't track position information:

```
nix-repl> x.json
{ foo = "bar"; }

nix-repl> builtins.unsafeGetAttrPos "foo" x.json
null
```

If there is a way to "ask" the library where the key is defined, it would not be difficult to convey that information back to Nix.

Thanks,
Graham

**Comment by nlohmann** [comment, marked as answer]
No, this is currently not possible. The lexer knows the position of each token, but this information is only used in case of parse errors.

**Reply by grahamc** [reply]
Thanks for the reply, would there be any interest in being able to access this information, or is that out of the scope of the library?

**Reply by nlohmann** [reply]
Why not - I think there was at least one similar request in the past. In any case, this should be an optional feature such that nobody needs to "pay" for source information. There already is a PR #3028 to store meta information. This could be one approach. Alternatively, it could be realized like the extended diagnostic information which add a member variable if `JSON_DIAGNOSTICS` is defined.

**Reply by barcode** [reply]
Well i wanted to add exactly this feature.
That was the reason why i wanted to be able to store meta information.

@nlohmann I started implementing it (https://github.com/barcode/json/tree/additional_information_for_sax_parser) and have two questions concerning this:
1)
I would go and extend the sax parser to optionally get this information. E.g. if the sax parser implements `null(const lexer&)` then it is called with the lexer. In case of bson (etc.) it would be possible to have `null(std::size_t pos)`. In the parser functions the calls to sax would be passed through a wrapper.
Do you think this is a good way to implement it?
2)
Also, this requires quiet a bit code (~50 lines per function (e.g. `is_null`)). This code is basically duplicated for each of the sax functions. (only difference are a few parameters)
I could use a preprocessor macro to avoid code duplication or avoid it by duplicating code.
What do you prefer? (some projects allow macros for such use cases, but i do not know what your policy is.)

/update
I finished the implementation and all existing tests are successful (so it looks i did not break anything). 
I still need to add tests actually using this feature.
Regarding 2): I used the version without a macro and `is_sax.hpp` (contains most impl code) went from 150 to ~~940~~ 760 (i will reduce this further) lines.

**Reply by barcode** [reply]
See https://github.com/nlohmann/json/pull/3165
